coreneuron/nrniv/nrn_acc_manager.cpp

```diff
@@ -42,371 +43,465 @@ void setup_nrnthreads_on_device(NrnThread* threads, int nthreads) {
+            }
+        }
+    }
 
-        dptr = d__data + 1 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_d), &(dptr), sizeof(double*));
+#else
+        if (nthreads <= 0) {
+            printf("\n Warning: No threads to copy on GPU! ");
+            return;
+        }
 
-        dptr = d__data + 2 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_a), &(dptr), sizeof(double*));
+        /** @todo: currently only checking nvidia gpu */
+        int num_gpus = acc_get_num_devices(acc_device_nvidia);
+        if (num_gpus == 0) {
+            printf("\n WARNING: Enabled GPU execution but couldn't find NVIDIA GPU! \n");
+        }
 
-        dptr = d__data + 3 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_b), &(dptr), sizeof(double*));
+        int i;
+        NrnThread* d_threads;
 
-        dptr = d__data + 4 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_v), &(dptr), sizeof(double*));
+        /* @todo: why dt is not setup at this moment? */
+        for (i = 0; i < nthreads; i++) {
+            (threads + i)->_dt = dt;
+            /* this thread will be computed on GPU */
+            (threads + i)->compute_gpu = 1;
+        }
 
-        dptr = d__data + 5 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_area), &(dptr), sizeof(double*));
+        /* -- copy NrnThread to device. this needs to be contigious vector because offset is used to
+         * find
+         * corresponding NrnThread using Point_process in NET_RECEIVE block
+         */
+        d_threads = (NrnThread*)acc_copyin(threads, sizeof(NrnThread) * nthreads);
 
-        if (nt->_actual_diam) {
-            dptr = d__data + 6 * ne;
-            acc_memcpy_to_device(&(d_nt->_actual_diam), &(dptr), sizeof(double*));
+        if (interleave_info == NULL) {
+            printf("\n Warning: No permutation data? Required for linear algebra!");
         }
 
-        int* d_v_parent_index = (int*)acc_copyin(nt->_v_parent_index, nt->end * sizeof(int));
-        acc_memcpy_to_device(&(d_nt->_v_parent_index), &(d_v_parent_index), sizeof(int*));
+        /* pointers for data struct on device, starting with d_ */
 
-        /* nt._ml_list is used in NET_RECEIVE block and should have valid membrane list id*/
-        Memb_list** d_ml_list =
-            (Memb_list**)acc_copyin(nt->_ml_list, n_memb_func * sizeof(Memb_list*));
-        acc_memcpy_to_device(&(d_nt->_ml_list), &(d_ml_list), sizeof(Memb_list**));
+        for (i = 0; i < nthreads; i++) {
+            NrnThread* nt = threads + i;      // NrnThread on host
+            NrnThread* d_nt = d_threads + i;  // NrnThread on device
 
-        /* -- copy NrnThreadMembList list ml to device -- */
+            if (nt->end <= 0) {
+                // this is an empty thread and could be only artificial
+                // cells. In this case nothing is executed on gpu.
+                nt->compute_gpu = 0;
+                continue;
+            }
```

Comment:
similar to previous comment : this should be common to unified and non-unified implementation.